src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c

```diff
@@ -0,0 +1,806 @@
+        doCleanup();
+        return FALSE;
+    }
+
+    pw.context = fp_pw_context_new(
+            fp_pw_thread_loop_get_loop(pw.loop),
+            NULL,
+            0
+    );
+
+    if (!pw.context) {
+        DEBUG_SCREENCAST("⚠ Could not create a pipewire context\n", NULL);
+        doCleanup();
+        return FALSE;
+    }
+
+    if (fp_pw_thread_loop_start(pw.loop) != 0) {
+        DEBUG_SCREENCAST("⚠ Could not start pipewire thread loop\n", NULL);
+        doCleanup();
+        return FALSE;
+    }
+
+    fp_pw_thread_loop_lock(pw.loop);
+
+    pw.core = fp_pw_context_connect_fd(
+            pw.context,
+            pw.pwFd,
+            NULL,
+            0
+    );
+
+    if (!pw.core) {
+        DEBUG_SCREENCAST("⚠ Could not create pipewire core\n", NULL);
+        goto fail;
+    }
+
+    pw_core_add_listener(pw.core, &pw.coreListener, &coreEvents, NULL);
+
+    for (int i = 0; i < screenSpace.screenCount; ++i) {
+        struct PwStreamData *data =
+                (struct PwStreamData*) malloc(sizeof (struct PwStreamData));
+        if (!data) {
+            ERR("failed to allocate memory\n");
+            goto fail;
+        }
+
+        memset(data, 0, sizeof (struct PwStreamData));
+
+        struct ScreenProps *screen = &screenSpace.screens[i];
+        screen->data = data;
+
+        DEBUG_SCREEN_PREFIX(screen, "@@@ adding screen %i\n", i);
+        if (checkScreen(i, requestedArea)) {
+            if (!connectStream(i)){
+                goto fail;
+            }
+        }
+        DEBUG_SCREEN_PREFIX(screen, "@@@ screen processed %i\n", i);
+    }
+
+    fp_pw_thread_loop_unlock(pw.loop);
+
+    return TRUE;
+
+    fail:
+        fp_pw_thread_loop_unlock(pw.loop);
+        doCleanup();
+        return FALSE;
+}
+
+static gboolean isAllDataReady() {
+    for (int i = 0; i < screenSpace.screenCount; ++i) {
+        if (!screenSpace.screens[i].shouldCapture) {
+            continue;
+        }
+        if (!screenSpace.screens[i].captureDataReady ) {
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+
+static void *pipewire_libhandle = NULL;
+//glib_version_2_68 false for gtk2, as it comes from gtk3_interface.c
+
+extern gboolean glib_version_2_68;
+
+#define LOAD_SYMBOL(fp_name, name) do {                             \
+    (fp_name) = dlsym(pipewire_libhandle, name);                    \
+    if (!(fp_name)) {                                               \
+       debug_screencast("⚠ %s:%i error loading dl_symbol %s\n",     \
+                        __func__, __LINE__, name);                  \
+       goto fail;                                                   \
+    }                                                               \
+} while(0);
+
+static gboolean loadSymbols() {
+    if (!glib_version_2_68) {
+        DEBUG_SCREENCAST("glib version 2.68+ required\n", NULL);
+        return FALSE;
+    }
+
+    pipewire_libhandle = dlopen("libpipewire-0.3.so.0",
```

Comment:
There's a useful macro for wrapping library names: `VERSIONED_JNI_LIB_NAME("pipewire-0.3", "0")`

Reply:
Updated.